examples/purchase-tester/src/main/java/com/revenuecat/purchasetester/MainActivity.kt

```diff
@@ -1,13 +1,25 @@
 package com.revenuecat.purchasetester
 
+import android.content.Intent
 import android.os.Bundle
 import androidx.appcompat.app.AppCompatActivity
+import com.revenuecat.purchases.ExperimentalPreviewRevenueCatPurchasesAPI
+import com.revenuecat.purchases.deeplinks.DeepLinkHandler
 import com.revenuecat.purchases_sample.R
 
+@OptIn(ExperimentalPreviewRevenueCatPurchasesAPI::class)
 class MainActivity : AppCompatActivity() {
 
     override fun onCreate(savedInstanceState: Bundle?) {
         super.onCreate(savedInstanceState)
         setContentView(R.layout.activity_main)
+        DeepLinkHandler.handleDeepLink(intent)
+    }
+
+    override fun onNewIntent(intent: Intent?) {
+        super.onNewIntent(intent)
+        if (intent != null) {
+            DeepLinkHandler.handleDeepLink(intent)
```

Comment:
Your understanding is correct! As for the reason, the developer needs to handle some cases like:
- If the token is expired, they probably want to display some kind of message indicating that a new link will be sent to their email
- If they want to hold the redemption to a certain point in the app's lifetime and maybe with a confirmation message.

Additionally, it will be easier to return a result for the redemption from the `startRedemption` than from the `handleDeepLink`, in case we support multiple types of deep links in the future.

If they actually don't want to handle any of those cases, they can just call `startRedemption()` immediately.

Reply:
I understand the scenarios we'd like to cover. I'm a bit concerned about the long chain of callbacks here. We have:
1. `onCreate()` / `onNewIntent()`, in which `handleDeepLink()` is called.
2. `handleWebPurchaseRedemption()`, in which `redeemWebPurchase()` is called.
3. `handleResult()`, containing the result of the redemption. 

It's hard to follow the flow of the code through all these callbacks, if a developer is trying to reason about their own web redemption code. It possibly involves going through multiple files to see what each of these callbacks is doing. 

I have 2 alternatives in mind, mostly intended to challenge your assumptions haha.

### Idea 1: no RedeemWebPurchaseListener

One idea I had is to remove `RedeemWebPurchaseListener.handleWebPurchaseRedemption()` from the equation. We could define an instance method, e.g. in `Purchases`, along the lines of: 

```kotlin
fun getWebPurchaseRedeemerOrNull(intent: Intent): RedeemWebPurchaseListener.WebPurchaseRedeemer?
```

This will return null if the intent is not a web purchase intent. This can then be used like so: 

```kotlin
override fun onNewIntent(intent: Intent) {
    Purchases.sharedInstance
        .getWebPurchaseRedeemerOrNull(intent)
        ?.redeemWebPurchase { result -> }
}
```

Devs can keep a hold of the `WebPurchaseRedeemer`, or the `Intent`, if they don't want to redeem the purchase right away. 

### Idea 2: no RedeemWebPurchaseListener & no WebPurchaseRedeemer

To simplify even further, we could remove the concept of a `WebPurchaseRedeemer`. We could define an instance method in `Purchases` along the lines of: 

```kotlin
fun redeemWebPurchase(intent: Intent, resultListener: ResultListener)
```

The `ResultListener` would return an error result if the intent is not a web purchase intent. (If devs want to check whether the intent is a web purchase intent before calling `redeemWebPurchase()`, we can add a method `isWebPurchaseRedemptionIntent()` that they can use.)

This can then be used like so: 

```kotlin
override fun onNewIntent(intent: Intent) {
    Purchases.sharedInstance.redeemWebPurchase(intent) { result -> }
}
```

In this case, if devs don't want to redeem right away, they can keep a hold of the `Intent`, and call `redeemWebPurchase()` whenever they want.

-------
  
  
Again, these are mostly intended to challenge your assumptions and see if there are opportunities to simplify. If we can get away with it, I like idea 2 best because it's by far the simplest and removes a lot of concepts for devs to understand.

Reply:
Those are very valid thoughts! So the main points with both of those approaches are IMO:
- The developer doesn't have a way to detect if there is a redemption link beforehand (unless they check the intent themselves which wouldn't be a great API IMO). This could be useful if the developer wants to display some UI like: "You are redeeming your web Purchase!" "Yes/No". Maybe not a needed use case for now, but just thinking about this case.
- These work better if we don't support pre-configuration caching, which as you said separately, we may consider not supporting for now.

Let's chat about this over Slack/in person 👍